### ISSUE TITLE
[BUG] No backpropagation after micro-batch-id ≥ 3 with MPI backend on Jetson Orin AGX

### ISSUE THREAD

**Issue by bhw2718** [microsoft/DeepSpeed#7492]
**Describe the bug**
I am trying to use DeepSpeed on two Jetson AGX Orin 64GB machines with the MPI backend.
Here is my earlier experience with compiling UCX, OpenMPI, and Torch, which I also discussed on the NVIDIA forum:
[https://forums.developer.nvidia.com/t/request-for-pytorch-wheel-with-mpi-backend-on-jetson-orin/340949/3](https://forums.developer.nvidia.com/t/request-for-pytorch-wheel-with-mpi-backend-on-jetson-orin/340949/3)

However, I encountered problems during training.
With the following `ds_config.json` configuration:

```json
"train_batch_size": 256,
"train_micro_batch_size_per_gpu": 256,
"gradient_accumulation_steps": 1,
```

And running the training with pipeline parallelism using `mpirun` as below:

```bash
mpirun \
  --allow-run-as-root \
  --hostfile hosts.txt \
  --mca plm_rsh_agent "ssh -p 22222" \
  --mca pml ucx \
  --mca osc ucx \
  --mca btl ^openib \
  -x UCX_TLS=sm,self,tcp,cuda_copy,cuda_ipc \
  -x UCX_NET_DEVICES=eno1 \
  -x PYTHONUNBUFFERED=1 \
  -x PYTHONHASHSEED=random \
  -x PYTHONIOENCODING=utf-8 \
  -x PYTHONFAULTHANDLER=1 \
  -np 2 \
deepspeed --launcher=openmpi train.py \
  --epochs=1 \
  --deepspeed_config=ds_config.json \
  -p 2 \
  --backend=mpi > 2025-08-15-ds-with-ompi-batch-256.log
```

I observed several issues:

1. Training sometimes becomes unstable, causing the training loss to remain high and not converge (this is the major issue).
2. Validation loss is also high, which I suspect might be due to a mistake on my part (minor issue).
3. To debug, I attached a layer hook to the model and found that for `micro-batch-id >= 3`, no backpropagation occurs, although forward propagation continues (as shown in the attached figure).

<img width="1589" height="790" alt="Image" src="https://github.com/user-attachments/assets/05634d47-79cd-4132-b166-0844e1fccdf2" />

This makes me wonder if this is the root cause of why training does not converge.

---

**To Reproduce**

1. Set up the environment as described in the NVIDIA forum.
2. Run the training command as shown above.
3. Observe CSV logs across different nodes.
4. Plot a Gantt chart.
5. You should see the same behavior (no backprop after micro-batch-id ≥ 3).

---

**ds\_report output**
See attached `ds_report.log`.

[ds_report.log](https://github.com/user-attachments/files/21793846/ds_report.log)

---

**System info**

* OS: Ubuntu 24.04.1 LTS (Noble Numbat)
* L4T version: 36.4.4
* Driver Version: 540.4.0
* CUDA Version: {'name': 'CUDA SDK', 'version': '12.8.1'}
* Python version: 3.12.3
* PyTorch version: 2.8.0a0+gitc525a02
* cmake version: 4.0.0
* gcc version: 13.3.0
* Hardware: 2 × Jetson Orin AGX (64GB)
* Both nodes connected under the same switch

---

**Launcher context**
Launched using `mpirun` as described above.

---

**Docker context**
I built a custom Docker image based on `dustynv/pytorch:2.7-r36.4.0-cu128-24.04` with the following main steps:

* Installed **UCX**, **OpenMPI**, and **PyTorch (custom bundle)** from precompiled tarballs.
* Installed **cusparselt** libraries from NVIDIA local repo.
* Installed required packages: `openssh-server`, `pdsh`, `iperf3`, `net-tools`, etc.
* Installed Python dependencies: `torch==2.8.0a0+gitc525a02`, `deepspeed`, `transformers`, `datasets`, `peft`, `scipy`, `tensorboard`.
* Configured SSH to run on port `22222` with root login enabled (for MPI cluster).
* Added startup script to generate SSH keys, run `ds_report`, and launch `sshd`.

---

**Additional context**
None.

---
Thanks a lot for your hard work on DeepSpeed, and I really appreciate any help or guidance on this issue.

---
[tensor_stats_rank0.csv](https://github.com/user-attachments/files/21793829/tensor_stats_rank0.csv): Layer hook results recorded for rank 0.
[tensor_stats_rank1.csv](https://github.com/user-attachments/files/21793828/tensor_stats_rank1.csv): Layer hook results recorded for rank 1.
[2025-08-15-ds-with-ompi-batch-256.log](https://github.com/user-attachments/files/21793830/2025-08-15-ds-with-ompi-batch-256.log): Training log output from the run.
[train.py](https://github.com/user-attachments/files/21793825/train.py): Training script
[debug_hooks.py](https://github.com/user-attachments/files/21793827/debug_hooks.py): The actual debug hook implementation.
[ds_config.json](https://github.com/user-attachments/files/21793826/ds_config.json): DeepSpeed configuration used in this run.
[hook_csv_parser.ipynb](https://github.com/user-attachments/files/21793916/hook_csv_parser.ipynb): Notebook for parsing the CSV hook results and visualizing them (used to generate the attached plots).
[Dockerfile.txt](https://github.com/user-attachments/files/21793903/Dockerfile.txt): Docker build recipe (renamed to .txt for GitHub upload; it is the actual Dockerfile).

**Comment by bhw2718** [comment]
**Update**

I’ve identified the root cause of the instability in my training runs, and it turned out to be an issue in my own code rather than DeepSpeed itself.

Specifically, I misunderstood how `engine.train_batch()` works with pipeline parallelism and gradient accumulation steps (GAS). Each call to `train_batch()` already consumes `GAS` micro-batches internally. In my earlier implementation I mistakenly iterated over `len(train_loader)` (which was already in units of micro-batches) without dividing by `GAS`. This caused the training loop to process `len(train_loader) * GAS` micro-batches per epoch, which made the learning rate scheduler and convergence behave inconsistently.

After correcting this so that the number of steps per epoch is computed as `ceil(len(train_loader) / GAS)`, training became much more stable.

Regarding the debugging hooks:
The reason I initially observed that backward propagation seemed to stop after micro-batch id ≥ 3 was simply that the model had entered the validation phase. During validation, only forward passes are executed, so no backward hooks are triggered. Once training resumes, backward activity continues as expected.

I’ll close this issue since the main problem is resolved.
